# Segmenting and Clustering Neighborhoods in Toronto

*GENERAL NOTE:  have hidden some output cells that are likely not relevant for grading this assignment; feel free to show them if necessary*

## Submittal 1

*Task:*  SCRAPE TORONTO POSTAL CODE DATA FROM A WEBSITE AND PLACE IN A DATAFRAME
--used two methods for this task as my 'old school approach' didn't seem 'worthy' of 10pts! 

**Method A:  Old School Approach** = copy data from website, clean data in Excel then import

*STEPS TAKEN OUTSIDE THIS NOTEBOOK*

    1. copy the data table from wikipedia page, paste into notepad, save as txt

    2. then import that into Excel and perform the desired manipulations in Excel; saved to sheet CleanedData

    a. remove rows with Not Assigned as the Borough

    b. replace / in Neighborhood with , using find/replace to match picture in instructions

    c. add Borough&PostalCode as Neighborhood where Neighborhood is blank -- this seems better (i.e. more explanatory) than just the borough by itself

    d. added Quay,South Niagara, Island airport back to M5V (in table this was a nested row!)

        NOTE:  didn't find any duplicate Postal Codes, even M5A was only listed once.

    3. loaded TorontoPostalCodes.xls into Skills Network Labs for use in this notebook


In [108]:
import pandas as pd

In [109]:
df= pd.read_excel('TorontoPostalCodes.xls')

In [110]:
df.head(12)

PostalCode           Borough  \
0         M4N   Central Toronto   
1         M4P   Central Toronto   
2         M4R   Central Toronto   
3         M4S   Central Toronto   
4         M4T   Central Toronto   
5         M4V   Central Toronto   
6         M5N   Central Toronto   
7         M5P   Central Toronto   
8         M5R   Central Toronto   
9         M4W  Downtown Toronto   
10        M4X  Downtown Toronto   
11        M4Y  Downtown Toronto   

                                         Neighborhood  
0                                       Lawrence Park  
1                                    Davisville North  
2                                  North Toronto West  
3                                          Davisville  
4                          Moore Park,Summerhill East  
5   Summerhill West,Rathnelly,South Hill,Forest Hi...  
6                                            Roselawn  
7                            Forest Hill North & West  
8                   The Annex,North Midtown,Yorkville  
9                                            Rosedale  
10                         St. James Town,Cabbagetown  
11                               Church and Wellesley

In [111]:
df.columns

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

In [112]:
df.shape

(103, 3)

so there are 103 assigned postal codes in the Toronto area!

**Method B = Scrape using BeautifulSoup** and then do all data wrangling in python (far more 'worthy' of 10pts!)

essentially followed tutorial at https://simpleanalytical.com/how-to-web-scrape-wikipedia-python-urllib-beautiful-soup-pandas

In [113]:
import urllib.request

In [114]:
#specify URL of website to scrape
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

# open the url using urllib.request and put the HTML into the page variable
page = urllib.request.urlopen(url)

In [115]:
!pip install beautifulsoup4 as bs4

In [116]:
# import the BeautifulSoup library so we can parse HTML and XML documents
from bs4 import BeautifulSoup

In [117]:
# parse the HTML from our URL into the BeautifulSoup parse tree format
soup = BeautifulSoup(page, "html.parser")

In [118]:
#print(soup.prettify())

In [119]:
# use the 'find_all' function to bring back all instances of the 'table' tag in the HTML and store in 'all_tables' variable
all_tables=soup.find_all("table")
#all_tables

In [120]:
right_table=soup.find('table', class_='wikitable sortable')
#right_table

In [121]:
#we want data from three columns of table, so we'll set up three empty lists
#use BS find all to locate string tr (table row), loop through all rows and in that loop use find all to locate string td (table data)
#add that data to the empty lists if the row of data has three elements
A=[]
B=[]
C=[]

for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)>=2:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

In [122]:
#use those lists to create dataframe
df2=pd.DataFrame(A,columns=['PostalCode'])
df2['Borough']=B
df2['Neighborhood']=C
df2.head()

PostalCode             Borough                  Neighborhood
0      M1A\n      Not assigned\n                            \n
1      M2A\n      Not assigned\n                            \n
2      M3A\n        North York\n                   Parkwoods\n
3      M4A\n        North York\n            Victoria Village\n
4      M5A\n  Downtown Toronto\n  Regent Park / Harbourfront\n

In [123]:
#hmm...looks like all cells have a \n in them.  let's look at the entire df2 to be sure
pd.set_option('display.max_rows', None)
df2

PostalCode             Borough  \
0        M1A\n      Not assigned\n   
1        M2A\n      Not assigned\n   
2        M3A\n        North York\n   
3        M4A\n        North York\n   
4        M5A\n  Downtown Toronto\n   
5        M6A\n        North York\n   
6        M7A\n  Downtown Toronto\n   
7        M8A\n      Not assigned\n   
8        M9A\n         Etobicoke\n   
9        M1B\n       Scarborough\n   
10       M2B\n      Not assigned\n   
11       M3B\n        North York\n   
12       M4B\n         East York\n   
13       M5B\n  Downtown Toronto\n   
14       M6B\n        North York\n   
15       M7B\n      Not assigned\n   
16       M8B\n      Not assigned\n   
17       M9B\n         Etobicoke\n   
18       M1C\n       Scarborough\n   
19       M2C\n      Not assigned\n   
20       M3C\n        North York\n   
21       M4C\n         East York\n   
22       M5C\n  Downtown Toronto\n   
23       M6C\n              York\n   
24       M7C\n      Not assigned\n   
25       M8C\n      Not assigned\n   
26       M9C\n         Etobicoke\n   
27       M1E\n       Scarborough\n   
28       M2E\n      Not assigned\n   
29       M3E\n      Not assigned\n   
30       M4E\n      East Toronto\n   
31       M5E\n  Downtown Toronto\n   
32       M6E\n              York\n   
33       M7E\n      Not assigned\n   
34       M8E\n      Not assigned\n   
35       M9E\n      Not assigned\n   
36       M1G\n       Scarborough\n   
37       M2G\n      Not assigned\n   
38       M3G\n      Not assigned\n   
39       M4G\n         East York\n   
40       M5G\n  Downtown Toronto\n   
41       M6G\n  Downtown Toronto\n   
42       M7G\n      Not assigned\n   
43       M8G\n      Not assigned\n   
44       M9G\n      Not assigned\n   
45       M1H\n       Scarborough\n   
46       M2H\n        North York\n   
47       M3H\n        North York\n   
48       M4H\n         East York\n   
49       M5H\n  Downtown Toronto\n   
50       M6H\n      West Toronto\n   
51       M7H\n      Not assigned\n   
52       M8H\n      Not assigned\n   
53       M9H\n      Not assigned\n   
54       M1J\n       Scarborough\n   
55       M2J\n        North York\n   
56       M3J\n        North York\n   
57       M4J\n         East York\n   
58       M5J\n  Downtown Toronto\n   
59       M6J\n      West Toronto\n   
60       M7J\n      Not assigned\n   
61       M8J\n      Not assigned\n   
62       M9J\n      Not assigned\n   
63       M1K\n       Scarborough\n   
64       M2K\n        North York\n   
65       M3K\n        North York\n   
66       M4K\n      East Toronto\n   
67       M5K\n  Downtown Toronto\n   
68       M6K\n      West Toronto\n   
69       M7K\n      Not assigned\n   
70       M8K\n      Not assigned\n   
71       M9K\n      Not assigned\n   
72       M1L\n       Scarborough\n   
73       M2L\n        North York\n   
74       M3L\n        North York\n   
75       M4L\n      East Toronto\n   
76       M5L\n  Downtown Toronto\n   
77       M6L\n        North York\n   
78       M7L\n      Not assigned\n   
79       M8L\n      Not assigned\n   
80       M9L\n        North York\n   
81       M1M\n       Scarborough\n   
82       M2M\n        North York\n   
83       M3M\n        North York\n   
84       M4M\n      East Toronto\n   
85       M5M\n        North York\n   
86       M6M\n              York\n   
87       M7M\n      Not assigned\n   
88       M8M\n      Not assigned\n   
89       M9M\n        North York\n   
90       M1N\n       Scarborough\n   
91       M2N\n        North York\n   
92       M3N\n        North York\n   
93       M4N\n   Central Toronto\n   
94       M5N\n   Central Toronto\n   
95       M6N\n              York\n   
96       M7N\n      Not assigned\n   
97       M8N\n      Not assigned\n   
98       M9N\n              York\n   
99       M1P\n       Scarborough\n   
100      M2P\n        North York\n   
101      M3P\n      Not assigned\n   
102      M4P\n   Central Toronto\n   
103      M5P\n   Central Toronto\n   
104      M6P\n

In [124]:
#convert to string so we can use str.replace 
#df2 = df2.astype(str)
df2 = df2.applymap(str)
df2.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [125]:
df2['PostalCode'] = df2['PostalCode'].str.replace('\n','')
df2['Borough'] = df2['Borough'].str.replace('\n','')
df2['Neighborhood'] = df2['Neighborhood'].str.replace('\n','')
df2.head()

PostalCode           Borough                Neighborhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

In [126]:
#want to drop rows with Borough = 'Not assigned'; let's see how many of those
df2['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East Toronto         5
East York            5
Mississauga          1
Name: Borough, dtype: int64

In [127]:
#so, we'll drop 77 rows 
df2.drop(df2.loc[df2['Borough']=='Not assigned'].index, inplace=True)
df2.head()

PostalCode           Borough                                  Neighborhood
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [128]:
# reset index
df2.reset_index(inplace=True)
del df2['index']
df2.head()

PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [129]:
#now let's replace all / in the Neighborhood column with commas.  e.g. this will change Regent Park / Harbourfront to Regent Park, Harbourfront
df2['Neighborhood'] = df2['Neighborhood'].str.replace(' / ',', ')
df2.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [130]:
df2.shape

(103, 3)

same result as in Method A, nice!

## Submittal 2

*Task:* GET LATITUDE & LONGITUDE DATA AND ADD TO DATAFRAME

**Method Shown: use provided .csv file**

    NOTE:  was NOT able to get geocoder to return lat/long coordinates at all.  was able to do so using OpenCage (another geocode provider)! -- in the interest of brevity and clarity, have only included the .csv steps here.

In [131]:
#url="https://cocl.us/Geospatial_data/Geospatial_Coordinates.csv"  got file url did not exist, so just imported that .csv into jupyter notebook
df_TorLatLong = pd.read_csv('Geospatial_Coordinates.csv')
df_TorLatLong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [132]:
df_TorLatLong.columns

Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')

In [133]:
#need to rename Postal Code column as PostalCode to match other dataframe so we can merge
df_TorLatLong.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)

In [134]:
df_TorLatLong.head(2)

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497

In [135]:
#let's make a copy of postal code dataframe, then switch the index to postal code
df3=df2
df3.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [136]:
df3.columns

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

In [137]:
df4=pd.merge(df3, df_TorLatLong, on="PostalCode")
df4.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

In [138]:
df4.shape

(103, 5)

same number of rows as in Submittal 1 with two more columns (latitude & longitude)...NICE!

## Submittal 3

*Task:* EXPLORE AND CLUSTER TORONTO NEIGHBORHOODS

First, let's import required libraries etc.

In [139]:
import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim 
import requests 
from pandas.io.json import json_normalize #not sure this is needed as we used pd.json_normalize in the lab to avoid errors
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library
print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


let's doublecheck the all Toronto neighborhoods data from above

In [140]:
df4.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Let's create a subset of Toronto neighborhoods for just those in Downtown Toronto

In [141]:
toronto_data = df4[df4['Borough']=='Downtown Toronto'].reset_index(drop=True)
print(toronto_data.shape)
toronto_data.head()

(19, 5)


PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

get Toronto geocoordinates and create map

In [142]:
#get lat/long of toronto for map creation
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Toronto geocoordinates are {}, {}.'.format(latitude, longitude))

Toronto geocoordinates are 43.6534817, -79.3839347.


In [143]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=13)

for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now let's get venue information for those Downtown Toronto neighborhoods

First, set FOURSQUARE credentials 

In [144]:
CLIENT_ID = 'LRIJEO12JGBDNEPKB42D1KASZJLER4UMNHGYJ4SYFAFMG0QK' 
CLIENT_SECRET = 'SQ1YGPKZJQFUKUK5KEX34531VLE2JOOVXB4QCDW34REW5MTD' 
VERSION = '20180605' 

In [145]:
#define function to get venues for many neighborhoods 
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [146]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [147]:
print(toronto_venues.shape)
toronto_venues.head()

(1215, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4       Breakfast Spot

In [148]:
toronto_venues.groupby('Neighborhood').count()
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 205 unique categories.


Now analyze each neighborhood, first split out the categories using *get_dummies*

In [149]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0         0            0                    0                 0   
1         0            0                    0                 0   
2         0            0                    0                 0   
3         0            0                    0                 0   
4         0            0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       1     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               1        0                0         0   

   Burger Joint  Burrito Place  Butcher  Cable Car  Café  Candy Store  \
0             0              0        0          0     0            0   
1             0              0        0          0     0            0   
2             0              0        0          0     0            0   
3             0              0        0          0     0            0   
4             0              0        0          0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0       

In [150]:
print(toronto_onehot.shape)
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
print(toronto_grouped.shape)
toronto_grouped.head()

(1215, 205)
(19, 205)


Neighborhood  Yoga Studio   Airport  \
0                                        Berczy Park     0.000000  0.000000   
1  CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.055556   
2                                 Central Bay Street     0.016949  0.000000   
3                                           Christie     0.000000  0.000000   
4                               Church and Wellesley     0.027397  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.055556      0.055556        0.111111         0.166667   
2            0.000000      0.000000        0.000000         0.000000   
3            0.000000      0.000000        0.000000         0.000000   
4            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000             0.000000           0.0       0.0     0.017857   
1          0.111111             0.000000           0.0       0.0     0.000000   
2          0.000000             0.000000           0.0       0.0     0.000000   
3          0.000000             0.000000           0.0       0.0     0.000000   
4          0.000000             0.013699           0.0       0.0     0.000000   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0             0.000000               0.0            0.000000   0.017857   
1             0.000000               0.0            0.000000   0.000000   
2             0.000000               0.0            0.000000   0.000000   
3             0.000000               0.0            0.055556   0.000000   
4             0.013699               0.0            0.000000   0.000000   

   Baby Store  Bagel Shop    Bakery  Bank       Bar  Baseball Stadium  \
0    0.000000    0.017857  0.035714   0.0  0.000000               0.0   
1    0.000000    0.000000  0.000000   0.0  0.055556               0.0   
2    0.000000    0.000000  0.000000   0.0  0.016949               0.0   
3    0.055556    0.000000  0.000000   0.0  0.000000               0.0   
4    0.000000    0.000000  0.000000   0.0  0.000000               0.0   

   Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0            0.017857  0.017857              0.0  0.035714         0.0   
1            0.000000  0.000000              0.0  0.000000         0.0   
2            0.000000  0.000000              0.0  0.000000         0.0   
3            0.000000  0.000000              0.0  0.000000         0.0   
4            0.000000  0.000000              0.0  0.013699         0.0   

   Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0  0.017857       0.000000   0.000000  0.000000   
1                 0.0  0.000000       0.055556   0.000000  0.055556   
2                 0.0  0.000000       0.000000   0.000000  0.000000   
3                 0.0  0.000000       0.000000   0.000000  0.000000   
4                 0.0  0.000000       0.000000   0.013699  0.000000   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                   0.0        0.017857      0.0         0.000000       0.0   
1                   0.0        0.000000      0.0         0.000000       0.0   
2                   0.0        0.000000      0.0         0.033898       0.0   
3                   0.0        0.000000      0.0         0.000000       0.0   
4                   0.0        0.013699      0.0         0.013699       0.0   

   Burger Joint  Burrito Place   Butcher  Cable Car      Café  Candy Store  \
0      0.000000       0.000000  0.017857   0.000000  0.035714     0.000000   
1      0.000000       0.000000  0.000000   0.000000  0.000000     0.000000   
2      0.033898       0.000000  0.000000   0.000000  0.050847     0.000000   
3      0.000000       0.000000  0.000000   0.000000  0.166667     0.055556   
4      0.027397   

In [151]:
#define function to sort venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [152]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar        Farmers Market   
1        Airport Lounge      Airport Terminal         Boat or Ferry   
2    Italian Restaurant                  Café        Sandwich Place   
3                  Café                  Park    Athletics & Sports   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery            Restaurant           Cheese Shop   
1       Harbor / Marina                   Bar                 Plane   
2           Salad Place       Bubble Tea Shop          Burger Joint   
3           Gas Station    Italian Restaurant                 Diner   
4                 Hotel                   Pub          Burger Joint   

  8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0    Italian Restaurant        Seafood Restaurant                   Café  
1   Rental Car Location               Coffee Shop       Sculpture Garden  
2        Ice Cream Shop       Japanese Restaurant             Comic Shop  
3             Nightclub               Candy Store             Restaurant  
4           Men's Store  Mediterranean Restaurant            Yoga Studio

CLUSTER NEIGHBORHOODS USING KMEANS

In [153]:
kclusters = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)
kmeans.labels_[0:10] 

array([0, 3, 4, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

*create new merged dataframe with clusters and venue info*

In [154]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = toronto_data
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head(20) # check the last columns!

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M5J  Downtown Toronto   
9         M5K  Downtown Toronto   
10        M5L  Downtown Toronto   
11        M5S  Downtown Toronto   
12        M5T  Downtown Toronto   
13        M5V  Downtown Toronto   
14        M4W  Downtown Toronto   
15        M5W  Downtown Toronto   
16        M4X  Downtown Toronto   
17        M5X  Downtown Toronto   
18        M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         Berczy Park  43.644771 -79.373306   
5                                  Central Bay Street  43.657952 -79.387383   
6                                            Christie  43.669542 -79.422564   
7                            Richmond, Adelaide, King  43.650571 -79.384568   
8   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
9            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
10                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
11                     University of Toronto, Harbord  43.662696 -79.400049   
12          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
13  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
14                                           Rosedale  43.679563 -79.377529   
15                                     Stn A PO Boxes  43.646435 -79.374846   
16                        St. James Town, Cabbagetown  43.667967 -79.367675   
17             First Canadian Place, Underground city  43.648429 -79.382280   
18                               Church and Wellesley  43.665860 -79.383160   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                0           Coffee Shop                Bakery   
1                4           Coffee Shop                 Diner   
2                0           Coffee Shop        Clothing Store   
3                0                  Café           Coffee Shop   
4                0           Coffee Shop          Cocktail Bar   
5                4           Coffee Shop    Italian Restaurant   
6                2         Grocery Store                  Café   
7                0           Coffee Shop                  Café   
8                0           Coffee Shop              Aquarium   
9                0           Coffee Shop                 Hotel   
10               0           Coffee Shop            Restaurant   
11               0                  Café    Italian Restaurant   
12               0                  Café           Coffee Shop   
13               3       Airport Service        Airport Lounge   
14               1                  Park                 Trail   
15               0           Coffee Shop    Italian Restaurant   
16               0           Coffee Shop    Italian Restaurant   
17               0           Coffee Shop                  Café   
18               0           Coffee Shop      Sushi Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                    Park                   Pub            Restaurant   
1        Sushi Restaurant                   Gym    Mexican Restaurant   
2                    Café        Cosmetics Shop   Japanese Restaurant   
3      Italian Restaurant             Gastropub          Cocktail Bar   
4                Beer Bar        Fa

In [155]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Here's an image of the map generated above stored at Github 
https://github.com/karldhuber/Coursera_Capstone/blob/master/DSCapstoneWeek3TorontoMAP.jpg 

EXAMINE & NAME CLUSTERS

Let's look at the top venues for each cluster and see if that allows us to come up with a cluster name.

Cluster 1

In [156]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                           Regent Park, Harbourfront               0   
2                            Garden District, Ryerson               0   
3                                      St. James Town               0   
4                                         Berczy Park               0   
7                            Richmond, Adelaide, King               0   
8   Harbourfront East, Union Station, Toronto Islands               0   
9            Toronto Dominion Centre, Design Exchange               0   
10                     Commerce Court, Victoria Hotel               0   
11                     University of Toronto, Harbord               0   
12          Kensington Market, Chinatown, Grange Park               0   
15                                     Stn A PO Boxes               0   
16                        St. James Town, Cabbagetown               0   
17             First Canadian Place, Underground city               0   
18                               Church and Wellesley               0   

   1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0            Coffee Shop                Bakery                   Park   
2            Coffee Shop        Clothing Store                   Café   
3                   Café           Coffee Shop     Italian Restaurant   
4            Coffee Shop          Cocktail Bar               Beer Bar   
7            Coffee Shop                  Café             Restaurant   
8            Coffee Shop              Aquarium     Italian Restaurant   
9            Coffee Shop                 Hotel                   Café   
10           Coffee Shop            Restaurant                   Café   
11                  Café    Italian Restaurant             Restaurant   
12                  Café           Coffee Shop  Vietnamese Restaurant   
15           Coffee Shop    Italian Restaurant                   Café   
16           Coffee Shop    Italian Restaurant                 Market   
17           Coffee Shop                  Café             Restaurant   
18           Coffee Shop      Sushi Restaurant    Japanese Restaurant   

   4th Most Common Venue 5th Most Common Venue      6th Most Common Venue  \
0                    Pub            Restaurant             Breakfast Spot   
2         Cosmetics Shop   Japanese Restaurant  Middle Eastern Restaurant   
3              Gastropub          Cocktail Bar        American Restaurant   
4         Farmers Market                Bakery                 Restaurant   
7                    Gym       Thai Restaurant                      Hotel   
8             Restaurant                  Café                      Hotel   
9             Restaurant   Japanese Restaurant        American Restaurant   
10                 Hotel                   Gym        American Restaurant   
11                   Bar   Japanese Restaurant                  Bookstore   
12          Dessert Shop    Mexican Restaurant              Grocery Store   
15    Seafood Restaurant                 Hotel                 Restaurant   
16            Restaurant                  Café                     Bakery   
17                 Hotel                   Gym                 Steakhouse   
18            Restaurant                 Hotel                        Pub   

   7th Most Common Venue 8th Most Common Venue     9th Most Common Venue  \
0                   Café               Theater                Shoe Store   
2        Bubble Tea Shop                 Diner                   Theater   
3                    Gym            Restaurant        Seafood Restaurant   
4            Cheese Shop    Italian Restaurant        Seafood Restaurant   
7          Deli / Bodega   American Restaurant                 Bookstore   
8    Sporting Goods Shop   Fried Chicken Joint            Scenic Lookout   
9            Salad Place    Seafood Restaurant       Sporting Goods Shop   
10    Seafood Restaurant    Italian Restaurant             Deli / Bodega   
11       

Cluster 2

In [157]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14     Rosedale               1                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14            Playground          Cupcake Shop            Donut Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14      Doner Restaurant               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
14        Discount Store                  Diner

Cluster 3

In [158]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6     Christie               2         Grocery Store                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                  Park    Athletics & Sports           Gas Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6    Italian Restaurant                 Diner             Nightclub   

  9th Most Common Venue 10th Most Common Venue  
6           Candy Store             Restaurant

Cluster 4

In [159]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
13  CN Tower, King and Spadina, Railway Lands, Har...               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
13       Airport Service        Airport Lounge      Airport Terminal   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
13         Boat or Ferry       Harbor / Marina                   Bar   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
13                 Plane   Rental Car Location           Coffee Shop   

   10th Most Common Venue  
13       Sculpture Garden

Cluster 5

In [160]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
1  Queen's Park, Ontario Provincial Government               4   
5                           Central Bay Street               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1           Coffee Shop                 Diner      Sushi Restaurant   
5           Coffee Shop    Italian Restaurant                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1                   Gym    Mexican Restaurant             Juice Bar   
5        Sandwich Place           Salad Place       Bubble Tea Shop   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1    Italian Restaurant            Hobby Shop   Fried Chicken Joint   
5          Burger Joint        Ice Cream Shop   Japanese Restaurant   

  10th Most Common Venue  
1    Distribution Center  
5             Comic Shop

NAMING ANALYSIS AND DISCUSSION

here's what I see in the venues for each cluster:
1. Cluster 1 has coffee shops, restaurants, cafes & hotels but no grocery stores or parks
1. Cluster 2 has park, dog run, playground, trail all listed highly no hotels nor cafes nor restaurants
1. Cluster 3 has grocery store listed first; also park, cafe
1. Cluster 4 has many airport and boat related venues.  so it must be an airport on an island!
1. Cluster 5 has coffee shops, diners, restaurants but no hotels.  so it's a business district without hotels probably right where the government buildings are (big area with no hotels allowed)  it's surrounded by Cluster 1 which has hotels

based on the above, here's what I've named the Downtown Toronto clusters:


**Cluster 1 = CENTRAL BUSINESS DISTRICT <br> Cluster 2 = PARK AREA <br> Cluster 3 = RESIDENTIAL AREA <br> Cluster 4 = AIRPORT ISLAND <br>Cluster 5 = GOVERNMENT DISTRICT**

AND THAT'S THE END OF THIS ASSIGNMENT!  MAYBE SOMEDAY I'LL GET TO VISIT TORONTO AND SEE THESE NEIGHBORHOODS!!